Tier-3 allowlist script failing

## :computer: System information

- Operating System: macOS
- Data Safe Haven version: 3.4.0+
- Powershell version: 7.2.5
- `Az` Powershell module version: 8.1.0
- `Microsoft.Graph` Powershell module version: 1.10.0
- `Poshstache` module version: 0.1.10

## :strawberry: Desired behaviour

The script which constructs tier-3 allowlists is failing

## :no_entry_sign: Actual behaviour

Allowlist build script is failing

```pwsh
Warning: 9 12:57:01 [WARNING]: No dependencies found for shapely (1.0) from pypi!
Get-Dependencies: /home/runner/work/data-safe-haven/data-safe-haven/deployment/administration/SHM_Expand_Allowlist_Dependencies.ps1:172
Line |
 172 |  … endencies = Get-Dependencies -Repository $Repository -Package $packag …
     |                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Cannot index into a null array.

Error: Process completed with exit code 1.
```

## :recycle: To reproduce

Look at recent CI jobs

--- a/deployment/administration/SHM_Expand_Allowlist_Dependencies.ps1
+++ b/deployment/administration/SHM_Expand_Allowlist_Dependencies.ps1
@@ -63,22 +63,30 @@ function Test-PackageExistence {
 function Get-Dependencies {
     param(
         [Parameter(Mandatory = $true, HelpMessage = "Name of package repository")]
+        [ValidateNotNullOrEmpty()]
         [string]$Repository,
         [Parameter(Mandatory = $true, HelpMessage = "Name of package to get dependencies for")]
+        [ValidateNotNullOrEmpty()]
         [string]$Package,
         [Parameter(Mandatory = $true, HelpMessage = "Versions of package to get dependencies for")]
-        $Versions,
+        [ValidateNotNullOrEmpty()]
+        [string[]]$Versions,
         [Parameter(Mandatory = $true, HelpMessage = "Hashtable containing cached dependencies")]
-        $Cache,
+        [ValidateNotNullOrEmpty()]
+        [hashtable]$Cache,
         [Parameter(Mandatory = $true, HelpMessage = "Only consider the most recent NVersions. Set to -1 to consider all versions.")]
+        [ValidateNotNullOrEmpty()]
         [int]$NVersions,
         [Parameter(Mandatory = $false, HelpMessage = "API key for libraries.io")]
         [string]$ApiKey
     )
     $dependencies = @()
-    if ($Package -notin $Cache[$Repository].Keys) { $Cache[$Repository][$Package] = [ordered]@{} }
     Add-LogMessage -Level Info "... found $($Versions.Count) versions of $Package"
     $MostRecentVersions = ($NVersions -gt 0) ? ($Versions | Select-Object -Last $NVersions) : $Versions
+    # If the package is not in the cache or is an invalid/empty object then replace it with a new ordered hashtable
+    if (($Package -notin $Cache[$Repository].Keys) -or (-not $Cache[$Repository][$Package])) {
+        $Cache[$Repository][$Package] = [ordered]@{}
+    }
     foreach ($Version in $MostRecentVersions) {
         if ($Version -notin $Cache[$Repository][$Package].Keys) {
             try {
@@ -101,6 +109,7 @@ function Get-Dependencies {
                 }
             } catch {
                 Add-LogMessage -Level Warning "No dependencies found for ${Package} (${Version}) from ${Repository}!"
+                $Cache[$Repository][$Package][$Version] = @()
             }
         }
         $dependencies += $Cache[$Repository][$Package][$Version]

